docs/dapp/sapphire/gasless.md

```diff
@@ -0,0 +1,305 @@
+in [EIP-155] format. These steps are executed as a confidential call. Finally,
+the user submits the generated transaction to the network.
+
+![Diagram of the On-Chain Signing](../images/sapphire/gasless-on-chain-signer.svg)
+
+### EIP155Signer
+
+To sign a transaction, the Sapphire's `EIP155Signer` library bundled along the
+`@oasisprotocol/sapphire-contract` package comes with the following helper which
+returns a raw, RLP-encoded, signed transaction ready to be broadcast:
+
+```solidity
+function sign(address publicAddress, bytes32 secretKey, EthTx memory transaction) internal view returns (bytes memory);
+```
+
+`publicAddress` and `secretKey` are the signer's address and their private key
+used to sign a meta-transaction (and pay for the fees). We will store these
+sensitive data inside the encrypted smart contract state together with the
+signer's `nonce` field in the following struct:
+
+```solidity
+struct EthereumKeypair {
+    address addr;
+    bytes32 secret;
+    uint64 nonce;
+}
+```
+
+The last `transaction` parameter in the `sign()` function is the transaction
+encoded in a format based on [EIP-155]. This can either be the original user's
+transaction or a meta-transaction.
+
+### Gasless Proxy Contract
+
+The following snippet is a complete *Gasless* contract for wrapping the user's
+transactions (`makeProxyTx()`) and executing them (`proxy()`). The signer's
+private key containing enough balance to cover transaction fees should be
+provided in the constructor.
+
+```solidity
+import {EIP155Signer} from '@oasisprotocol/sapphire-contracts/contracts/EIP155Signer.sol';
+
+struct EthereumKeypair {
+    address addr;
+    bytes32 secret;
+    uint64 nonce;
+}
+
+struct EthTx {
+    uint64 nonce;
+    uint256 gasPrice;
+    uint64 gasLimit;
+    address to;
+    uint256 value;
+    bytes data;
+    uint256 chainId;
+}
+
+contract Gasless {
+    EthereumKeypair private kp;
+
+    constructor (EthereumKeypair memory keypair)
+    payable
+    {
+        kp = keypair;
+    }
+    
+    function makeProxyTx(
+        uint256 gasPrice,
+        address innercallAddr,
+        bytes memory innercall
+    )
+    external view
+    returns (bytes memory output)
+    {
+        // Call will invoke proxy().
+        bytes memory data = abi.encode(innercallAddr, innercall);
+
+        return EIP155Signer.sign(kp.addr, kp.secret, EIP155Signer.EthTx({
+            nonce: kp.nonce,
+            gasPrice: gasPrice,
+            gasLimit: 250000,
+            to: address(this),
+            value: 0,
+            data: abi.encodeWithSelector(this.proxy.selector, data),
+            chainId: block.chainid
+        }));
+    }
+
+    function proxy(bytes memory data)
+    external payable
+    {
+        (address addr, bytes memory subcall_data) = abi.decode(data, (address, bytes));
+        (bool success,) = addr.call{value: msg.value}(subcall_data);
+        require( success );
```

Comment:
Maybe worth passing error messages through to caller, otherwise you will get obscure 'it failed' reverts without useful info.

```solidity
(success, out_data) = in_contract.call(in_data);
assembly {
    switch success
    case 0 { revert(add(out_data,32),mload(out_data))
}
```